File: compiler/nodes/types/tuple.go
package types

import (
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type TupleTypeNode struct {
	Elements []database.Node
	Facts    *database.Facts
}

func (node *TupleTypeNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseTupleType(parser *syntax.Parser) (*TupleTypeNode, *syntax.Error) {
	span := parser.Spanned()

	many, err := syntax.ParseMany(parser, 1, ParseTypeElement, func(parser *syntax.Parser) (*struct{}, *syntax.Error) {
		_, err := parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}

		parser.ConsumeLineBreaks()

		return &struct{}{}, nil
	})
	if err != nil {
		return nil, err
	}

	if len(many) == 1 {
		_, err = parser.Token("TupleOperator")
		if err != nil {
			return nil, err
		}
	} else {
		_, _, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) (struct{}, *syntax.Error) {
			_, err := parser.Token("TupleOperator")
			if err != nil {
				return struct{}{}, err
			}

			return struct{}{}, nil
		})
		if err != nil {
			return nil, err
		}
	}

	elements := make([]database.Node, 0, len(many))
	for _, item := range many {
		elements = append(elements, item.Value)
	}

	return &TupleTypeNode{
		Elements: elements,
		Facts:    database.NewFacts(span()),
	}, nil
}

func (node *TupleTypeNode) Visit(visitor *visit.Visitor) {
	visitType(visitor, node)

	for _, element := range node.Elements {
		visitor.Visit(element)
	}

	visitor.Constraint(typecheck.NewTypeConstraint(node, typecheck.TupleType(node.Elements)))
}

File: compiler/src/nodes/types/tuple.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{parse_type_element, visit_type},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct TupleTypeNode {
    pub elements: Vec<NodeRef>,
}

impl Node for TupleTypeNode {}

pub fn parse_tuple_type(parser: &mut Parser<'_>) -> Result<TupleTypeNode, ParseError> {
    let elements = parser
        .parse_many(1, parse_type_element, |parser| {
            parser.token(TokenKind::TupleOperator)?;
            parser.consume_line_breaks();
            Ok(())
        })?
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    if elements.len() == 1 {
        parser.token(TokenKind::TupleOperator)?;
    } else {
        let _ = parser.parse_optional(|parser| {
            parser.token(TokenKind::TupleOperator)?;
            Ok(())
        })?;
    }

    Ok(TupleTypeNode { elements })
}

impl Visit for TupleTypeNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_type(node, visitor);

        for element in &self.elements {
            visitor.visit(element);
            visitor.edge(element, node, "element");
        }

        visitor.constraint(TypeConstraint::new(
            node.clone(),
            visitor.tuple_type(self.elements.iter().cloned()),
        ));
    }
}

impl Codegen for TupleTypeNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Err(ctx.error())
    }
}
